cpp/src/arrow/util/ubsan.h

```diff
@@ -63,7 +63,7 @@ inline std::enable_if_t<std::is_trivially_copyable_v<T>, T> SafeLoadAs(
 template <typename T>
 inline std::enable_if_t<std::is_trivially_copyable_v<T>, T> SafeLoad(const T* unaligned) {
   std::remove_const_t<T> ret;
-  std::memcpy(&ret, unaligned, sizeof(T));
+  std::memcpy(&ret, reinterpret_cast<const void*>(unaligned), sizeof(T));
```

Comment:
This should probably just be static cast

Reply:
What is the difference exactly in this case? Is `reinterpret_cast` stricter, or the converse?

Reply:
static cast is more strict and therefore preferable here; reinterpret cast should only be used when the type system is being circumvented somehow (eg [type punning](https://github.com/apache/arrow/blob/main/cpp/src/arrow/util/aligned_storage.h#L36))

The [legal operations for reinterpret cast](https://timsong-cpp.github.io/cppwp/n4868/expr.reinterpret.cast) don't include many of the things we use it for, but ... the type punning we do in arrow-c++ is not unique to this project.

Reply:
In C it's undefined behavior to **materialize** an unaligned pointer whereas in C++ it's only undefined to access such. The new assertion in ubsan is a nice example of compiler optimizations encroaching into previously "safe" undefined behavior; since memcpy's parameter is `void const*` it initially appears that memcpy shouldn't care about the pointer's alignment. However, since the compiler knows that we're accessing the pointer and can therefore assume that it points to aligned memory... the memcpy call can always be replaced by an aligned load.

In light of this, I think we probably want to throw a `std::launder` in here as well- otherwise in the future, the compiler could trace through the cast and make the same breaking assumption again
```suggestion
  std::memcpy(&ret, std::launder(static_cast<const void*>(unaligned)), sizeof(T));
```

Reply:
Hmm, the doc for `std::launder` says:

> Devirtualization fence with respect to p. Returns a pointer to the same memory that p points to, but where the referent object is assumed to have a distinct lifetime and dynamic type. 

Does the "distinct lifetime" imply that ASAN wouldn't be able to detect use-after-free?

Reply:
Also, this does not seem to be one of the "typical uses of std::launder" as mentioned in https://en.cppreference.com/w/cpp/utility/launder

Reply:
I think I'd rather merge without `std::launder` for now. We can reevaluate later if the issue arises, IMHO.